rfd/0139-gogo-must-go.md

```diff
@@ -0,0 +1,279 @@
+custom JSON and YAML marshalers for our objects. This may be time intensive but not overall
+a significant amount of work.
+
+#### Embedding types (removing `gogoproto.embed`)
+
+We heavily use `gogoproto.embed` for generating objects with a consistent structure or interface.
+We particularly use these with our `kind`, `version`, and `metadata` fields. This will be
+difficult to mitigate without some heavy lifting. As such, I recommend the custom storage object
+format listed above and the addition of `FromV*` functions in order to convert protobuf objects
+into this object. For example:
+
+```go
+  accessRequest := accessrequest.FromV1Proto(request.AccessRequest)
+  user := user.FromV6Proto(request.User)
+```
+
+#### Casting types (removing `gogoproto.casttype`)
+
+We additionally rely heavily on `gogoproto.casttype`, which allows us to avoid doing explicit
+casts or conversions within go code. We will have to take on this conversion work, unfortunately.
+
+#### Other gogoproto tags
+
+##### `gogoproto.nullable`
+
+We will need to remove the `gogoproto.nullable` tags, which overall shouldn't be a huge impact.
+We will need to rely on our `CheckAndSetDefaults` functions, which should generally handle this
+today.
+
+##### `gogoproto.stdtime`
+
+We will need to migrate to using `timestamppb` directly instead of relying on gogo's `stdtime` tag.
+
+### Proposed solution
+
+#### Create internal objects
+
+Currently, in `api/types` we define a large number of interfaces that are subsequently used to wrap
+our protobuf created messages in a nicer, more user friendly way, abstracting away much of the protobuf
+bits and pieces.
+
+We should break this relationship and create actual internal objects to replace interfaces. From here, we
+should use conversion functions to translate between protobuf messages and our internal objects.
+
+It's imperative that these objects are compatible with the current JSON/YAML representations in Teleport.
+That is, the objects that are in Teleport today should unmarshal properly into our new internal objects.
+
+All objects should remain in `api/types` as they do today. (Open question: should this be the case? Should
+we relocate objects?)
+
+##### Example
+
+Let's take `UserGroup` as an example, as it's a very simple message. `UserGroup`'s interface as it is
+currently defined in `api/types`:
+
+```go
+// UserGroup specifies an externally sourced group.
+type UserGroup interface {
+	ResourceWithLabels
+
+	// GetApplications will return a list of application IDs associated with the user group.
+	GetApplications() []string
+	// SetApplications will set the list of application IDs associated with the user group.
+	SetApplications([]string)
+}
+```
+
+For this message, I propose we replace the above interface with a struct, along
+with a builder for convenience and avoiding breakages:
+
+```go
+type UserGroupBuilder struct {
+  applications []string
+}
```

Comment:
Why a builder compared to just having a New function that does the same thing from a spec?

Reply:
A builder would be useful, IMO, because it allows us to change `New*` functions without breaking `e` builds. Other than that, however, I don't feel too strongly.

Reply:
We could start with a `New(spec)` function by default, and add the builder for more complex scenarios. As long as breaking the builds doesn't become a habit we can be "optimistic" with the initial declarations.

Reply:
I'll remove references to the builder.